jax/_src/scipy/signal.py

```diff
@@ -146,3 +151,341 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
       coef, *_ = linalg.lstsq(A, data[sl])
       data = data.at[sl].add(-jnp.matmul(A, coef, precision=lax.Precision.HIGHEST))
     return jnp.moveaxis(data.reshape(shape), 0, axis)
+
+
+def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides):
+  """Calculate windowed FFT in the same way the original SciPy does.
+  """
+  # Created strided array of data segments
+  if nperseg == 1 and noverlap == 0:
+    result = x[..., np.newaxis]
+  else:
+    step = nperseg - noverlap
+    *batch_shape, signal_length = x.shape
+    batch_shape = tuple(batch_shape)
+    x = x.reshape((int(np.prod(batch_shape)), signal_length))[..., np.newaxis]
+    result = jax.lax.conv_general_dilated_patches(
+        x, (nperseg,), (step,),
+        'VALID',
+        dimension_numbers=('NTC', 'OIT', 'NTC'))
+    result = result.reshape(batch_shape + result.shape[-2:])
+
+  # Detrend each data segment individually
+  result = detrend_func(result)
+
+  # Apply window by multiplication
+  result = win.reshape((1,) * len(batch_shape) + (1, nperseg)) * result
+
+  # Perform the fft on last axis. Zero-pads automatically
+  if sides == 'twosided':
+    return jax.numpy.fft.fft(result, n=nfft)
+  else:
+    return jax.numpy.fft.rfft(result.real, n=nfft)
+
+
+def odd_ext(x, n, axis=-1):
+  """Extends `x` along with `axis` by odd-extension.
+
+  This function was previously a part of "scipy.signal.signaltools" but is no
+  longer exposed.
+
+  Args:
+    x : input array
+    n : the number of points to be added to the both end
+    axis: the axis to be extended
+  """
+  if n < 1:
+    return x
+  if n > x.shape[axis] - 1:
+    raise ValueError(
+        f"The extension length n ({n}) is too big. "
+        f"It must not exceed x.shape[axis]-1, which is {x.shape[axis] - 1}.")
+  left_end = lax.slice_in_dim(x, start_index=0, limit_index=1, axis=axis)
```

Comment:
Nit: looking at the JAX codebase, people usually call `lax.slice_in_dim` with `start_index` and `limit_index` as positional arguments. Maybe you could do the same here.

Reply:
done.